src/box/memtx_tx.c

```diff
@@ -762,74 +762,33 @@ memtx_tx_adjust_position_in_read_view_list(struct txn *txn)
 	rlist_add_tail(&prev_txn->in_read_view_txs, &txn->in_read_view_txs);
 }
 
-/**
- * Mark @a victim as conflicted and abort it.
- * Does nothing if the transaction is already aborted.
- */
-static void
-memtx_tx_abort_with_conflict(struct txn *victim)
+void
+memtx_tx_abort_with_conflict(struct txn *txn)
 {
-	assert(victim->psn == 0);
-	if (victim->status == TXN_ABORTED)
-		return;
-	if (victim->status == TXN_IN_READ_VIEW)
-		rlist_del(&victim->in_read_view_txs);
-	victim->status = TXN_ABORTED;
-	txn_set_flags(victim, TXN_IS_CONFLICTED);
-}
-
-/**
- * Handle conflict when @a victim has read and @a breaker has written the same
- * key, and @a breaker is prepared. The functions must be called in two cases:
- * 1. @a breaker becomes prepared for every victim with non-empty intersection
- * of victim read set / breaker write set.
- * 2. @a victim has to read confirmed value and skips the value that prepared
- * @a breaker wrote.
- * If @a victim is read-only or hasn't made any changes, it should be sent
- * to read view, in which is will not see @a breaker's changes. If @a victim
- * is already in a read view - a read view that does not see every breaker
- * changes is chosen.
- * Otherwise @a victim must be marked as conflicted and aborted on occasion.
- *
- * NB: can trigger story garbage collection.
- */
-static void
-memtx_tx_handle_conflict(struct txn *breaker, struct txn *victim)
-{
-	assert(breaker != victim);
-	assert(breaker->psn != 0);
-	assert(victim->psn == 0);
-	if (victim->status != TXN_INPROGRESS &&
-	    victim->status != TXN_IN_READ_VIEW) {
-		/* Was conflicted by somebody else. */
-		return;
-	}
-	if (stailq_empty(&victim->stmts)) {
-		assert((victim->status == TXN_IN_READ_VIEW) ==
-		       (victim->rv_psn != 0));
-		/* Send to read view, perhaps a deeper one. */
-		if (victim->status != TXN_IN_READ_VIEW) {
-			victim->status = TXN_IN_READ_VIEW;
-			victim->rv_psn = breaker->psn;
-			rlist_add_tail(&txm.read_view_txs,
-				       &victim->in_read_view_txs);
-		} else if (victim->rv_psn > breaker->psn) {
-			/*
-			 * Note that in every case for every key we may choose
-			 * any read view psn between confirmed level and the
-			 * oldest prepared transaction that changes that key.
-			 * But we choose the latest level because it generally
-			 * costs less, and if there are several breakers - we
-			 * must sequentially decrease read view level.
-			 */
-			victim->rv_psn = breaker->psn;
-			assert(victim->rv_psn != 0);
-		}
-		memtx_tx_adjust_position_in_read_view_list(victim);
-	} else {
-		/* Mark as conflicted. */
-		memtx_tx_abort_with_conflict(victim);
+	if (txn->status == TXN_IN_READ_VIEW)
+		rlist_del(&txn->in_read_view_txs);
+}
+
+void
+memtx_tx_send_to_read_view(struct txn *txn, int64_t psn)
+{
+	assert((txn->status == TXN_IN_READ_VIEW) == (txn->rv_psn != 0));
+	if (txn->status != TXN_IN_READ_VIEW) {
+		txn->status = TXN_IN_READ_VIEW;
+		txn->rv_psn = psn;
+		rlist_add_tail(&txm.read_view_txs, &txn->in_read_view_txs);
```

Comment:
Consider the case:
1. Your transaction is started only in `vinyl`. It is in read-view with psn = 100.
2. You read from a `memtx` space - transaction there is started, and it should have the same read-view psn 100.

The problem here is that memtx MVCC GC could delete too old stories, so it has stories only with psn > 200 - memtx read-view would be inconsistent in this case.
I believe that vinyl MVCC will face the same problem, but don't know for sure.

It seems there is a simple solution - we could make the `txm.read_view_txs` list public and all engines could take the lowest read-view psn from it.

Reply:
Note that `txn_send_to_read_view` starts the transaction in all engines that support MVCC before sending it to a read view so the situation when one engine (say vinyl) is in a read view while another (say memtx) isn't is impossible.